frontend/middle-2/design.md

```diff
@@ -1,17 +1,46 @@
 #Базовые принципы проектирования
 
-* Что такое инкапсуляция? Какова мотивация для улучшения уровня инкапсуляции?
-* Что такое абстрагирование? Какие есть отличия между абстракцией, инкапсуляцией и скрытием информации?
-* Что такое coupling? Как инкапсуляция влияет на это?
-* Что такое cohesion? Является ли эта величина обратной величине coupling?
-* Что такое принцип DRY? Как принцип связан с cohesion?
-* Что такое принцип KISS?
-* Что такое принцип YAGNI? Как он соотносится с принципом KISS?
-* Что такое Separation of concerns?
-* Из-за чего возникает проблема impedance mismatch при обмене данными? Какие есть примеры возникновения проблемы в разных технологиях? Какие там используются решения?
-* Почему преждевременная оптимизация может легко обернуться проблемой, а не решением при проектировании систем?
-
+* Inversion of Control
+  * What is an IoC?
+  * What are advantages of framework with IoC over libraries?
+  * What are advantages of libraries without IoC over frameworks?
+  * Can we combine approach of using IoC and avoiding its within interface of the same module?
+* Dependency Injection
+  * What is DI?
+  * What are DI alternatives? What are advantages and disadvantages of each alternative?
+  * Can we implement DI without IoC?
+  * How DI help us to achieve the Separation of Concerns?
+* What is the Impedance Mismatch? What are examples of Impedance Mismatch among different technologies? What are most popular patterns of solutions?
+* Why premature optimization can be easily turn into big problems? How to avoid infecting the system by problems of premature optimization?
+* What are benefits to expose the API with different level of abstraction: from high level to lower ones?
+* Why callback break composability?
+* Large Scale JS
+  * What is the main idea of the proposed approach?
+  * What patterns are used in this approach?
+  * What are role of events in this approach? Can we implement the similar approach only with synchronous function invocations?
+  * How does it help to achieve fault tolerance?
+  * How modules can depend on each other?
+* API of modules
+  * What are characteristics of good API of the module?
+  * How can we make our API hard to misuse? What are benefits of this characteristic?
+  * What are consistency and conceptual integrity of the API? Why they are valuable properties?
```

Comment:
What are consistency and conceptual integrity properties of the API? Why are the properties valuable?